### PR TITLE
add editableTextId - a stable unique TextEditable identifier

### PR DESCRIPTION
**Pull request by yjbanov**
Attach an `editableTextId` value to the `EditableTextState` that uniquely identifies the text field and pass it to the engine via text input channel and semantics nodes. This field will be used to fix a race between the `flutter/textinput` channel and semantics.

This makes progress towards fixing https://github.com/flutter/flutter/issues/52106 (engine changes are required to fully fix this).

## Background

On the web, the semantics tree is responsible for hosting the `<input>` elements. The semantics update happens after the `TextInputConnection` is established. Conversely, the semantics tree can create `<input>` elements before text input connection is established. This leads to the following bad situations leading to various bugs:

- A text input command is sent to the engine, but there's no `<input>` element to sync to.
- Text input command is routed to a wrong `<input>` element.
- `<input>` is removed (due to lazy rendering) while a text input connection is still alive.

## Solution

`editableTextId` outlives `clientId` (which is unique to a single `TextInputConfiguration`), and it is the single source of truth for both semantics and the text input channel. Whether one or the other initiates the creation of an `<input>` element, they can connect by matching on the `editableTextId`.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -2327,6 +2329,17 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     }
   }
 
+  int get editableTextId => _editableTextId;
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:2332]: here and elsewhere: missing docs

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 2]
```diff
@@ -206,6 +206,7 @@ class SemanticsData with Diagnosticable {
     this.tags,
     this.transform,
     this.customSemanticsActionIds,
+    required this.editableTextId,
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart:209]: Should this only be set if the flags indicate that it is a textfield? Maybe assert?

File: packages/flutter/lib/src/rendering/editable.dart [hunk 3]
```diff
@@ -2327,6 +2329,17 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
...
+  int _editableTextId;
+  set editableTextId(int value) {
+    assert(value != null);
+    if (value != _editableTextId) {
+      _editableTextId = value;
+      markNeedsPaint();
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/editable.dart:2338]: Why does this require a paint update (seems to be only used in describeSemanticsConfiguration)?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 4]
```diff
@@ -3907,6 +3929,11 @@ class SemanticsConfiguration {
   /// configurations as determined by [isCompatibleWith].
   void absorb(SemanticsConfiguration child) {
     assert(!explicitChildNodes);
+    assert(
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart:3932]: Is this condition already ensured by isCompatibleWith? isCompatibleWith should probably mark configurations that have different _editableTextIds as incompatible?

**Comment by goderbauer** [comment]
> editableTextId outlives clientId (which is unique to a single TextInputConfiguration)

Instead of inventing a new ID could we just make clientID live longer and use that?

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/rendering/editable.dart:2332]: Yeah, lots of clean up to do, lots of docs to write. For now just sending a draft. Also scheduled a VC for this Thursday to discuss if we're OK with this approach overall.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 2 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart:209]: Unfortunately, const asserts don't support the expression `flag & SemanticsFlag.isTextField.value == 0`. If you can think of a workaround, I'm all ears. FWIW, I'll have this assertion in `dart:ui`.

File: packages/flutter/lib/src/rendering/editable.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter/lib/src/rendering/editable.dart:2338]: Done. For some reason I thought a semantics change always implied a repaint, but I guess that doesn't have to be the case.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 4 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart:3932]: Done. I would also like to assert that no two editables are nested. Is there a good place for such an assertion?

Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart:3932]: Although, I suppose a single text field can be described by two nodes each supplying a different subset of semantic properties, so maybe we should allow it.

**Comment by yjbanov** [comment]
> > editableTextId outlives clientId (which is unique to a single TextInputConfiguration)
> 
> Instead of inventing a new ID could we just make clientID live longer and use that?

If we are sure that `TextInputConnection` never needs a unique ID, then I can give it a shot. It may be a breaking change for platform implementors/embedders due to the expectation that once you send `TextInput.clearClient` its clientID is no longer valid and all resources associated with it can be freed. It would also be mildly breaking on the framework side as the client ID would have to move from the hidden `TextInputConnection._id` field to, say, a public `TextInputClient.id` field. Other than that, I don't see why not.

**Comment by yjbanov** [comment]
Closing in favor of an approach that doesn't require framework-side changes (https://github.com/flutter/engine/pull/25797)

**Pull request closed without merging**